powerauth-push-server/src/main/java/io/getlime/push/service/DeviceRegistrationService.java

```diff
@@ -0,0 +1,262 @@
+/*
+ * Copyright 2024 Wultra s.r.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.getlime.push.service;
+
+import com.wultra.security.powerauth.client.PowerAuthClient;
+import com.wultra.security.powerauth.client.model.enumeration.ActivationStatus;
+import com.wultra.security.powerauth.client.model.error.PowerAuthClientException;
+import com.wultra.security.powerauth.client.model.response.GetActivationStatusResponse;
+import io.getlime.push.errorhandling.exceptions.PushServerException;
+import io.getlime.push.model.enumeration.MobilePlatform;
+import io.getlime.push.model.request.CreateDeviceForActivationsRequest;
+import io.getlime.push.model.request.CreateDeviceRequest;
+import io.getlime.push.model.request.UpdateDeviceStatusRequest;
+import io.getlime.push.repository.PushDeviceRepository;
+import io.getlime.push.repository.model.AppCredentialsEntity;
+import io.getlime.push.repository.model.Platform;
+import io.getlime.push.repository.model.PushDeviceRegistrationEntity;
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.dao.DataIntegrityViolationException;
+import org.springframework.retry.annotation.Backoff;
+import org.springframework.retry.annotation.Retryable;
+import org.springframework.stereotype.Component;
+
+import java.util.Date;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Service managing registration and persistence of push devices.
+ *
+ * @author Jan Pesek, jan.pesek@wultra.com
+ */
+@Component
+@AllArgsConstructor
+@Slf4j
+public class DeviceRegistrationService {
+
+    private final PushDeviceRepository pushDeviceRepository;
+    private final PowerAuthClient powerAuthClient;
+
+    @Retryable(retryFor = DataIntegrityViolationException.class,
+            maxAttemptsExpression = "${powerauth.push.service.registration.retry.maxAttempts:2}",
+            backoff = @Backoff(delayExpression = "${powerauth.push.service.registration.retry.backoff:100}"))
+    public void createOrUpdateDevice(final CreateDeviceRequest requestObject, final AppCredentialsEntity appCredentials) throws PushServerException {
```

Comment:
Should we start the transaction here? I realize that this glitch used to be already there.